src/main/java/org/zalando/nakadi/service/ConsumerLimitingService.java

```diff
@@ -0,0 +1,231 @@
+                // if slots for all partitions were acquired - the connection is successful
+                return slots.stream()
+                        .map(Optional::get)
+                        .collect(toList());
+            } else {
+                // if a slot for at least one partition wasn't acquired - connection can't be created
+                final String failedPartitions = partitions.stream()
+                        .filter(p -> !slots.stream().anyMatch(s -> s.isPresent() && s.get().getPartition().equals(p)))
+                        .collect(joining(", "));
+                final String msg = format(ERROR_MSG, eventType, failedPartitions, maxConnections);
+                throw new NoConnectionSlotsException(msg);
+            }
+        } catch (final Exception e) {
+            // in a case of failure release slots for partitions that already acquired slots
+            slots.stream()
+                    .filter(Optional::isPresent)
+                    .map(Optional::get)
+                    .forEach(this::releaseConnectionSlot);
+            if (e instanceof NoConnectionSlotsException) {
+                throw e;
+            } else {
+                throw new ServiceUnavailableException("Error communicating with zookeeper", e);
+            }
+        }
+    }
+
+    public void releaseConnectionSlots(final List<ConnectionSlot> connectionSlots) {
+        connectionSlots.forEach(this::releaseConnectionSlot);
+    }
+
+    private void releaseConnectionSlot(final ConnectionSlot slot) {
+        final String consumerNode = zkNodeNameForConsumer(slot.getClient(), slot.getEventType(), slot.getPartition());
+        final String connectionNodePath = format("{0}/{1}/{2}",
+                CONNECTIONS_ZK_PATH, consumerNode, slot.getConnectionId());
+        try {
+            zkHolder.get()
+                    .delete()
+                    .guaranteed()
+                    .forPath(connectionNodePath);
+            deletePartitionNodeIfPossible(consumerNode);
+        } catch (final Exception e) {
+            LOG.error("Zookeeper error when deleting consumer connection node", e);
+        }
+
+        ACQUIRED_SLOTS.remove(slot);
+        try {
+            deleteCacheIfPossible(slot);
+        } catch (final Exception e) {
+            LOG.error("Zookeeper error when deleting consumer connections cache", e);
+        }
+    }
+
+    private void deleteCacheIfPossible(final ConnectionSlot slot) throws IOException {
+        final boolean hasMoreConnectionsToPartition = ACQUIRED_SLOTS.stream()
+                .anyMatch(s -> s.getPartition().equals(slot.getPartition())
+                        && s.getClient().equals(slot.getClient())
+                        && s.getEventType().equals(slot.getEventType()));
+        if (!hasMoreConnectionsToPartition) {
+            final String consumerPath = zkPathForConsumer(slot.getClient(), slot.getEventType(), slot.getPartition());
```

Comment:
In general, I did not get why you need ACQUIRED_SLOTS. You can get acquired slots from cache because ACQUIRED_SLOTS do not give you any synchronisation between cache and ZK.